File: core/java/src/main/java/hara/lang/protocol/IStreamPoll.java
package hara.lang.protocol;

import hara.lang.declaration.HaraAvailability;
import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

/** Non-blocking readable stream capability. */
@HaraProtocolBinding(
    namespace = "std.protocol.istreampoll",
    name = "IStreamPoll",
    availability = HaraAvailability.CAPABILITY_GATED,
    capability = "native-runtime-protocols")
public interface IStreamPoll {
  @HaraMethod(value = "poll", arity = 1)
  Object poll();
}

File: core/rust/src/lang/protocol/istreampoll.rs
use hara_protocol_macros::hara_protocol;

#[hara_protocol(
    namespace = "std.protocol.istreampoll",
    name = "IStreamPoll",
    availability = "capability-gated",
    capability = "native-runtime-protocols"
)]
pub trait IStreamPoll {
    type Item;

    #[hara_method(value = "poll", arity = 1)]
    fn poll(&mut self) -> Option<Self::Item>;
}
